hazelcast/src/main/java/com/hazelcast/sql/impl/worker/QueryOperationWorker.java

```diff
@@ -0,0 +1,172 @@
+
+import com.hazelcast.internal.nio.Packet;
+import com.hazelcast.internal.serialization.SerializationService;
+import com.hazelcast.internal.util.concurrent.MPSCQueue;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.sql.HazelcastSqlException;
+import com.hazelcast.sql.impl.QueryId;
+import com.hazelcast.sql.impl.QueryUtils;
+import com.hazelcast.sql.impl.operation.QueryCancelOperation;
+import com.hazelcast.sql.impl.operation.QueryOperation;
+import com.hazelcast.sql.impl.operation.QueryOperationDeserializationException;
+import com.hazelcast.sql.impl.operation.QueryOperationHandler;
+
+import java.util.UUID;
+
+import static com.hazelcast.instance.impl.OutOfMemoryErrorDispatcher.inspectOutOfMemoryError;
+import static com.hazelcast.sql.impl.QueryUtils.WORKER_TYPE_OPERATION;
+
+/**
+ * Worker responsible for operation processing.
+ */
+public class QueryOperationWorker implements Runnable {
+
+    private static final Object POISON = new Object();
+
+    private final QueryOperationHandler operationHandler;
+    private final SerializationService ss;
+    private final Thread thread;
+    private final MPSCQueue<Object> queue;
+    private final ILogger logger;
+
+    private UUID localMemberId;
+
+    public QueryOperationWorker(
+        QueryOperationHandler operationHandler,
+        SerializationService ss,
+        String instanceName,
+        int index,
+        ILogger logger
+    ) {
+        this.operationHandler = operationHandler;
+        this.ss = ss;
+        this.logger = logger;
+
+        thread = new Thread(this,  QueryUtils.workerName(instanceName, WORKER_TYPE_OPERATION, index));
+        queue = new MPSCQueue<>(thread, null);
+
+        thread.start();
+    }
+
+    public void init(UUID localMemberId) {
+        this.localMemberId = localMemberId;
+    }
+
+    public void submit(QueryOperationExecutable task) {
+        queue.add(task);
+    }
+
+    public void stop() {
+        queue.clear();
+        queue.add(POISON);
+
+        thread.interrupt();
+    }
+
+    @Override
+    public void run() {
+        try {
+            run0();
+        } catch (Throwable t) {
+            inspectOutOfMemoryError(t);
+            logger.severe(t);
+        }
+    }
+
+    private void run0() {
+        try {
+            while (true) {
+                Object task = queue.take();
+
+                if (task == POISON) {
+                    break;
+                } else {
+                    assert task instanceof QueryOperationExecutable;
+
+                    execute((QueryOperationExecutable) task);
+                }
+            }
+        } catch (InterruptedException e) {
+            // No-op.
+        }
+    }
+
+    private void execute(QueryOperationExecutable task) {
+        QueryOperation operation = task.getLocalOperation();
+
+        if (operation == null) {
+            operation = deserialize(task.getRemoteOperation());
+        }
+
+        operationHandler.execute(operation);
```

Comment:
It's possible to pass `null` to the `execute`. Is it allowed? Is it documented anywhere?

Reply:
It should not be possible. Fixed the code.